crypto/ec/ec_asn1.c

```diff
@@ -792,18 +794,79 @@ EC_GROUP *EC_GROUP_new_from_ecparameters(const ECPARAMETERS *params)
         goto err;
     }
 
+    /*
+     * Check if the explicit parameters group just created matches one of the
+     * built-in curves.
+     *
+     * We create a copy of the group just built, so that we can remove optional
+     * fields for the lookup: we do this to avoid that one of the optional
+     * parameters is used to force the library into using a less performant and
+     * less secure EC_METHOD instead of the specialized one.
+     * In any case, `seed` is not really used in any computation, while a
+     * cofactor different from the one in the built-in table is just
+     * mathematically wrong anyway and should not be used.
+     */
+    if (NULL == (ctx = BN_CTX_new())) {
+        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_BN_LIB);
+        goto err;
+    }
+    if (NULL == (dup = EC_GROUP_dup(ret))
+            || 1 != EC_GROUP_set_seed(dup, NULL, 0)
+            || !EC_GROUP_set_generator(dup, point, a, NULL)) {
+        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_EC_LIB);
+        goto err;
+    }
+    if (NID_undef != (curve_name = ec_curve_nid_from_params(dup, ctx))) {
+        /*
+         * The input explicit parameters successfully matched one of the
+         * built-in curves: often for built-in curves we have specialized
+         * methods with better performance and hardening.
+         *
+         * In this case we replace the `EC_GROUP` created through explicit
+         * parameters with one created from a named group.
+         */
+        EC_GROUP *named_group = NULL;
+
+#ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
+        /*
+         * NID_wap_wsg_idm_ecid_wtls12 and NID_secp224r1 are both aliases for
+         * the same curve, we prefer the SECP nid when matching explicit
+         * parameters as that is associated with a specialized EC_METHOD.
+         */
+        if (curve_name == NID_wap_wsg_idm_ecid_wtls12)
+            curve_name = NID_secp224r1;
+#endif /* !def(OPENSSL_NO_EC_NISTP_64_GCC_128) */
+
+        if (NULL == (named_group = EC_GROUP_new_by_curve_name(curve_name))) {
+            ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_EC_LIB);
+            goto err;
+        }
+        EC_GROUP_clear_free(ret);
```

Comment:
Why do we need to do `EC_GROUP_clear_free()` rather than just `EC_GROUP_free()`. (I also wonder why that function exists at all?)

Reply:
I was actually thinking about this: it does not make much sense to me either, but the original code used it and I thought changing it internally and deprecating `EC_GROUP_clear_free()` should go in a separate PR.

Do you feel strongly about this? I can change just limited to this function if you prefer!

Reply:
I will fix the occurrences in this file, and open a separate issue to track the removal from the rest of the code and possibly deprecation.

Reply:
Fixed by a7875052d4

Reply:
Opened #9822 to track the rest.

Reply:
Works for me